torch/utils/checkpoint.py

```diff
@@ -335,7 +340,370 @@ def forward(input):
         )
     return run_function(end + 1, len(functions) - 1, functions)(input)
 
-def _checkpoint_without_reentrant(function, preserve_rng_state=True, *args, **kwargs):
+# TODO: this doc needs to be updated!
+#
+# NOTE [ Nestable Checkpoint ]
+#
+# The semantics of nested checkpoint can be defined by two basic rules.
+# Following the two rules leads to an important implication that is central
+# to motivating the design.
+#
+# Rule 1. Saved tensors are managed by inner-most checkpoint only and hidden
+#         from any outer layers of checkpoint.
+#
+# Rule 2. The inputs of inner checkpoints are treated as tensors saved to its
+#         parent checkpoint.
+#
+# Implication: To recompute any given saved tensor, we need to recompute all of
+#              the checkpoints wrapping it.
+#
+# Why is this implied? To unpack a saved tensor X during backward we need to
+# recompute the inner-most checkpoint (#1), and in order to recompute that
+# checkpoint I need to have its inputs, which are managed by that checkpoint's
+# parent (#2), which thus also needs to be recomputed first. Continue this line
+# of reasoning and we realize that in order to unpack X, all checkpoints that
+# were active at the time X was saved need to be recomputed. (unless we have
+# already done so in that backward for some other saved tensor).
+#
+# In practice, we use a noop autograd Function to save inputs as saved tensors.
+# During unpack calling ctx.saved_tensor triggers the parent checkpoint to
+# recompute.
+#
+# [ Implementation details of Rule 2 ]
+#
+# Rule 2 specifies that we should store the inputs of inner checkpoints onto the
+# parent, but this is not possible if the checkpoint is already the outer-most.
+# In practice, we consider two cases:
+#
+#  1) If checkpoint does not have a parent, we store them directly on the
```

Comment:
Yeah not necessary, but I think we should keep it anyway as an optimization since vast majority of cases are non-nested. This optimization is nice especially if it is as disruptive as it is now
(will update docs to reflect this)

Reply:
Well, if that breaks save_on_cpu(checkpoint(fn)), then I prefer not to keep it ;)